### PR TITLE
Show more detail in browse image streams page

### LINKED ISSUE #2987: Web Console should show image streams with no status tags
If `status.tags` are missing, image streams don't show up in the UI at all. It is possible for the "No image steams" message to disappear entirely on the browse page, leaving it just empty whitespace.

See #2224
See #2961

### PR DESCRIPTION
**Pull request by spadgett**
Fixes #2987

### CODE REVIEW

**Comment by spadgett** [comment]
Before:

openshift_web_console

After:

openshift_web_console

With image stream tag tags:

openshift_web_console

With more status.tag.items:

openshift_web_console

openshift_web_console

/cc @bparees

**Comment by bparees** [comment]
lgtm.

File: assets/app/views/images.html [hunk 1]
```diff
@@ -10,30 +10,44 @@
...
-          <div>
-            <span>{{imageStream.metadata.name}}:{{tag.tag}}</span>
-            <span ng-if="!tag.items.length"> &mdash; <em>No images for this tag</em></span>
-            <span ng-if="tag.items.length && tag.items[0].image">
-              <span> &mdash; </span>
-              <short-id id="{{tag.items[0].image | imageName}}"></short-id>
-              <span> created </span>
-              <relative-timestamp timestamp="tag.items[0].created"></relative-timestamp>
+      <div ng-repeat="imageStream in imageStreams" class="tile">
+        <h2>{{imageStream.metadata.name}}
```
Comment (liggitt) [inline assets/app/views/images.html]: newlines before content makes structure easier to see, e.g.
```
<h2>
  {{imageStream.metadata.name}}
  <small class="tile-details">
    Created <relative-timestamp...
```

**Comment by liggitt** [comment]
I think we should defer this

**Comment by spadgett** [comment]
Rebased and made some updates after feedback from @liggitt . Still working on calling out spec tags that don't have corresponding status tags.

@liggitt @bparees thoughts? A bit more compact now.

openshift_web_console

**Comment by bparees** [comment]
@spadgett works for me.

**Comment by liggitt** [comment]
is the mismatch between the tag and the pull spec just due to sample data you loaded?

**Comment by spadgett** [comment]
It's the data.

```
      "spec": {
        "dockerImageRepository": "openshift/mongodb-24-centos7",
        "tags": [
          {
            "name": "2.4",
            "from": {
              "kind": "ImageStreamTag",
              "name": "latest"
            }
          },
          {
            "name": "latest"
          }
        ]
      }
```

...

```
          {
            "tag": "2.4",
            "items": [
              {
                "created": "2015-06-25T19:45:48Z",
                "dockerImageReference": "openshift/mongodb-24-centos7:latest",
                "image": "2f6a8f1bcbb43b1830888fb724605f5a3c8d3b5190cb46dc73a4098a6ea778f4"
              }
            ]
          },
          {
            "tag": "latest",
            "items": [
              {
                "created": "2015-06-25T19:45:48Z",
                "dockerImageReference": "openshift/mongodb-24-centos7:latest",
                "image": "2f6a8f1bcbb43b1830888fb724605f5a3c8d3b5190cb46dc73a4098a6ea778f4"
              }
            ]
          }
```

**Comment by spadgett** [comment]
Spec tag with no corresponding status tag:

openshift_web_console

**Comment by spadgett** [comment]
Rebased

File: assets/app/scripts/controllers/images.js [hunk 2]
```diff
@@ -23,10 +24,37 @@ angular.module('openshiftConsole')
       LabelFilter.setLabelSuggestions($scope.labelSuggestions);
       $scope.imageStreams = LabelFilter.getLabelSelector().select($scope.unfilteredImageStreams);
       $scope.emptyMessage = "No image streams to show";
+      updateMissingStatusTags();
       updateFilterWarning();
       Logger.log("image streams (subscribe)", $scope.imageStreams);
     }));
 
+    // Check each image stream to see if the spec tags resolve to status tags.
```
Comment (liggitt) [inline assets/app/scripts/controllers/images.js]: isn't this backwards? I thought we cared about image streams with status tags that weren't resolved to spec tags

Comment (liggitt) [inline assets/app/scripts/controllers/images.js]: nm, I got it backwards

File: assets/app/views/images.html [hunk 3]
```diff
@@ -10,30 +10,48 @@
           <em>{{emptyMessage}}</em>
         </div>
       </div>
-      <div ng-repeat="imageStream in imageStreams">
-        <div class="tile" ng-repeat="tag in imageStream.status.tags">
+      <div ng-repeat="imageStream in imageStreams track by (imageStream | uid)" class="tile">
+        <h2>
+          {{imageStream.metadata.name}}
+          <small class="tile-details">Created
+            <relative-timestamp timestamp="imageStream.metadata.creationTimestamp"></relative-timestamp></small>
```
Comment (liggitt) [inline assets/app/views/images.html]: is the timestamp the image stream was created important?

File: assets/app/views/images.html [hunk 4]
```diff
@@ -10,30 +10,48 @@
...
-              <short-id id="{{tag.items[0].image | imageName}}"></short-id>
-              <span> created </span>
-              <relative-timestamp timestamp="tag.items[0].created"></relative-timestamp>
+            {{specTag.name}}
+            &mdash;
+            <span class="pficon-layered">
+              <span class="pficon pficon-warning-triangle"></span>
+              <span class="pficon pficon-warning-exclamation"></span>
             </span>
+            Tag could not be resolved.
```
Comment (liggitt) [inline assets/app/views/images.html]: @ncdc is there any info recorded in the status for the UI to tell the difference between a sync not happening yet and a sync failing?

Comment (ncdc) [inline assets/app/views/images.html]: There's an annotation in the stream that will either have a time stamp
(success) or error text. I can show you tomorrow if you'd like.

Comment (spadgett) [inline assets/app/views/images.html]: > There's an annotation in the stream that will either have a time stamp (success) or error text. I can show you tomorrow if you'd like.

@ncdc yes, that would help. thanks

File: assets/app/views/images.html (hunk 3 above)
Comment (spadgett) [inline assets/app/views/images.html]: Hm, maybe not. I can take it out.

File: assets/app/views/images.html (hunk 4 above)
Comment (spadgett) [inline assets/app/views/images.html]: Here's the annotation:

```
        "annotations": {
          "openshift.io/image.dockerRepositoryCheck": "the repository \"openshift/ruby-999-centos7\" was not found"
        }
```

On success:

```
        "annotations": {
          "openshift.io/image.dockerRepositoryCheck": "2015-07-17T18:38:55Z"
        }
```

**Comment by spadgett** [comment]
Removed the creation date for the image stream itself.

Still syncing:

screen_shot_2015-07-17_at_3_23_18_pm

Missing status tags:

openshift_web_console

The message from the annotation is available when hovering over the warning icon.

**Comment by liggitt** [comment]
The warning icon seems like it should be after the tag name, and the same size as the "could not be resolved" text. Then all the tags align left.

**Comment by spadgett** [comment]
I'm OK either way, but I thought it looked odd that the warning icons didn't line up.

File: assets/app/views/images.html [hunk 5]
```diff
@@ -10,30 +10,53 @@
...
         </div>
       </div>
-      <div ng-repeat="imageStream in imageStreams">
-        <div class="tile" ng-repeat="tag in imageStream.status.tags">
+      <div ng-repeat="imageStream in imageStreams track by (imageStream | uid)" class="tile">
+        <h2>
+          {{imageStream.metadata.name}}
+        </h2>
+        <div ng-if="imageStream.spec.dockerImageRepository" class="small">Docker image repository: {{imageStream.spec.dockerImageRepository}}</div>
+        <div ng-if="!(imageStream | annotation : 'openshift.io/image.dockerRepositoryCheck') && !imageStream.status.tags.length"
```
Comment (liggitt) [inline assets/app/views/images.html]: If the sync hasn't happened yet, should we still show the status tags, but put the "not synced" text after each one?

Comment (spadgett) [inline assets/app/views/images.html]: Yeah, I think that's better.

screen_shot_2015-07-17_at_4_21_32_pm

**Comment by spadgett** [comment]
Updated to show each status tag with not synced message. I haven't changed the icon position for now because I like how it looks...

File: assets/app/views/images.html [hunk 6]
```diff
@@ -10,30 +10,55 @@
...
+              <span class="pficon pficon-warning-exclamation"></span>
             </span>
+            {{specTag.name}}
+            &mdash;
+            <span class="small">Tag could not be resolved.</span>
+          </div>
+          <div ng-if="!(imageStream | annotation : 'openshift.io/image.dockerRepositoryCheck')">
+              {{specTag.name}}
+              &mdash;
+              <span class="small text-muted">Not synced.</span>
```
Comment (liggitt) [inline assets/app/views/images.html]: `Not yet synced`? No period (it's a fragment), and "yet" implies that it should happen without their intervention

**Comment by spadgett** [comment]
Should we be able to collapse image streams with many tags? Or status tags that have no spec tags? For instance, here is the official node repo with twice as many tags off screen.

openshift_web_console

**Comment by spadgett** [comment]
@liggitt updated message to "Not yet synced"

File: assets/app/styles/_tile.less [hunk 7]
```diff
@@ -14,6 +14,11 @@
   h1, h2, h3 {
     margin: (@line-height-computed / 2) 0px;
   }
+  .tile-details {
```
Comment (jwforres) [inline assets/app/styles/_tile.less]: did you check everywhere we use tile-details to make sure it doesn't cause any regressions?

File: assets/app/views/images.html [hunk 8]
```diff
@@ -10,30 +10,55 @@
...
+              <span ng-if="tag.items.length && tag.items[0].image">&mdash;
+                <short-id id="{{tag.items[0].image | imageName}}"></short-id>
+                <small class="text-muted">Created <relative-timestamp timestamp="tag.items[0].created"></relative-timestamp></small>
+              </span>
+            <span ng-if="!tag.items.length"><em>No images for this tag</em></span>
+          </div>
+          <div class="small">
+            <!-- dockerImageReference values can get long. Use max width and class truncate to elide some of the SHA1 for long values.
+                 They can still be copied even when elided. -->
+            <div ng-if="tag.items.length && tag.items[0].dockerImageReference" style="max-width: 400px;" class="truncate"
```
Comment (jwforres) [inline assets/app/views/images.html]: does this work still on the smallest mobile

**Comment by jwforres** [comment]
@spadgett for the image streams with a lot of tags, i agree on hiding some.  Maybe show the first 10 and then do a 'See all (50)...' or something similar

File: assets/app/styles/_tile.less (hunk 7 above)
Comment (spadgett) [inline assets/app/styles/_tile.less]: It's a new style not used anywhere else. (I searched to be sure.)

File: assets/app/views/images.html (hunk 8 above)
Comment (spadgett) [inline assets/app/views/images.html]: Here is 300px:

openshift_web_console

Comment (jwforres) [inline assets/app/views/images.html]: k seems fine to me

**Comment by liggitt** [comment]
LGTM, Jessica has final ok/merge

**Comment by jwforres** [comment]
LGTM @spadgett why dont you open a separate issue to track collapsing when there are a lot of them.  Are you ready for merge on what is here?

**Comment by jwforres** [comment]
[merge]

**Comment by spadgett** [comment]
```
==> openshiftdev: Downloading artifacts from '/tmp/origin/e2e/' to '/var/lib/jenkins/jobs/merge_pull_requests_origin/workspace/origin/artifacts/e2e/'
Warning: Permanently added '54.80.140.79' (RSA) to the list of known hosts.
Address 54.80.140.79 maps to ec2-54-80-140-79.compute-1.amazonaws.com, but this does not map back to the address - POSSIBLE BREAK-IN ATTEMPT!
receiving incremental file list
rsync: change_dir "/tmp/origin/e2e" failed: No such file or directory (2)
```

```
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1505) [receiver=3.0.6]
==> openshiftdev: Unable to download artifacts
Build step 'Execute shell' marked build as failure
```

**Comment by jwforres** [comment]
[merge]

**Comment by jwforres** [comment]
and again [merge]

**Pull request merged**